File: src/main/java/com/kingmang/lazurite/parser/AST/Statements/PrintStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.console.Console;
import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public final class PrintStatement extends InterruptableNode implements Statement {
    
    public final Expression expression;
    @Override
    public void execute() {
        super.interruptionCheck();
        Console.print(expression.eval().asString());
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return "print " + expression;
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/PrintStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.console.Console
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor

data class PrintStatement(val expression: Expression) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        Console.print(this.expression.eval().asString())
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        "print ${this.expression}"
}
